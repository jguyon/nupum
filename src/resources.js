import invariant from "tiny-invariant";
import { createResource } from "./resource-cache";

const NPMS_API = "https://api.npms.io/v2";

export const packageSearch = createResource(
  async ({ query, from = 0, size = 25 }) => {
    invariant(typeof query === "string", "expected query to be a string");
    invariant(typeof from === "number", "expected from to be a number");
    invariant(typeof size === "number", "expected size to be a number");

    const params = new URLSearchParams({
      q: query,
      from,
      size,
    });
    const response = await fetch(`${NPMS_API}/search?${params.toString()}`);

    invariant(
      response.status === 200,
      `expected response status to be 200 but got ${response.status}`,
    );

    return await response.json();
  },
  ({ query, from = 0, size = 25 }) => JSON.stringify([query, from, size]),
);

export const packageInfo = createResource(async name => {
  invariant(typeof name === "string", "expected name to be a string");

  const response = await fetch(
    `${NPMS_API}/package/${encodeURIComponent(name)}`,
  );

  if (response.status === 200) {
    const data = await response.json();

    if (
      !data.collected.metadata.readme &&
      data.collected.metadata.links &&
      data.collected.metadata.links.repository
    ) {
      data.collected.metadata.readme = await fetchGithubReadme(
        data.collected.metadata.links.repository,
      );
    }

    return {
      found: true,
      data,
    };
  } else if (response.status === 404) {
    return {
      found: false,
    };
  } else {
    invariant(
      false,
      `expected response status to be 200 or 404 but got ${response.status}`,
    );
  }
});

async function fetchGithubReadme(repository) {
  let matches;
  if ((matches = repository.match(/^https:\/\/github.com\/([^/]+)\/([^/]+)/))) {
    const org = matches[1];
    const repo = matches[2];

    const response = await fetch(
      `https://raw.githubusercontent.com/${org}/${repo}/master/README.md`,
    );

    if (response.status === 200) {
      return await response.text();
    } else {
      return undefined;
    }
  } else {
    return undefined;
  }
}
