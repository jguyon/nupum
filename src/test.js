import React, { StrictMode } from "react";
import { render as rtlRender, cleanup } from "react-testing-library";
import "jest-dom/extend-expect";
import serializer from "jest-emotion";

expect.addSnapshotSerializer(serializer);

afterEach(() => {
  cleanup();
});

export * from "react-testing-library";

export function render(ui, options) {
  const { rerender: rtlRerender, ...rest } = rtlRender(wrapUi(ui));

  function rerender(nextUi) {
    rtlRerender(wrapUi(nextUi));
  }

  return {
    ...rest,
    rerender,
  };
}

function wrapUi(ui) {
  return <StrictMode>{ui}</StrictMode>;
}
