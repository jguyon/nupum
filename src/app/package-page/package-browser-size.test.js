import React from "react";
import { renderWithContext, act } from "../../test/react";
import { fakePackageBrowserSize } from "../../test/fake";
import { packageBrowserSize } from "../../resources";
import PackageBrowserSize from "./package-browser-size";

test("loading indicator is rendered while request fetches", () => {
  const { container } = renderWithContext(
    <PackageBrowserSize name="react" version="16.8.6" />,
  );

  expect(container).toMatchSnapshot();
});

test("error indicator is rendered when request fails", () => {
  const { resourceCache, container } = renderWithContext(
    <PackageBrowserSize name="react" version="16.8.6" />,
  );

  act(() => {
    resourceCache.fail(packageBrowserSize, {
      name: "react",
      version: "16.8.6",
    });
  });

  expect(container).toMatchSnapshot();
});

test("browser size is rendered when request succeeds", () => {
  const { resourceCache, container } = renderWithContext(
    <PackageBrowserSize name="react" version="16.8.6" />,
  );

  act(() => {
    resourceCache.succeed(
      packageBrowserSize,
      { name: "react", version: "16.8.6" },
      fakePackageBrowserSize({ name: "react", version: "16.8.6" }),
    );
  });

  expect(container).toMatchSnapshot();
});
