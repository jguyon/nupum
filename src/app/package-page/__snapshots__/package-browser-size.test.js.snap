// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`browser size is rendered when request succeeds 1`] = `
<div>
  2.1KB
</div>
`;

exports[`error indicator is rendered when request fails 1`] = `
<div>
  N/A
</div>
`;

exports[`loading indicator is rendered while request fetches 1`] = `
.emotion-0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<div>
  <span
    class="emotion-0"
  >
    Loading
  </span>
  â€¦
</div>
`;
